### PR TITLE
Improve pause logic

### LINKED ISSUE #2694: Gameplay cursor can get in an incorrect clicked state
Something about overlays appearing in front of it while in a "pressed" state will get it stuck in said state.

https://puu.sh/Awsq5/0bf52f429e.mp4

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/59068093-gameplay-cursor-can-get-in-an-incorrect-clicked-state?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #3077: Mania/taiko gameplay input still gets propagated while paused
Hitsounds play and if you happen to pause on a note, can hit the note by pressing the key.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/61112716-mania-taiko-gameplay-input-still-gets-propagated-while-paused?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by peppy**
- Closes #2694.
- Fixes #3077

---

### CODE REVIEW

File: osu.Game/Screens/Play/Player.cs [hunk 1]
```diff
@@ -166,12 +166,8 @@ private void load(AudioManager audio, APIAccess api, OsuConfigManager config)
...
                     OnQuit = Exit,
                     CheckCanPause = () => AllowPause && ValidForResume && !HasFailed && !RulesetContainer.HasReplayLoaded,
-                    OnPause = () =>
-                    {
-                        pauseContainer.Retries = RestartCount;
-                        hudOverlay.KeyCounter.IsCounting = !pauseContainer.IsPaused;
-                    },
-                    OnResume = () => hudOverlay.KeyCounter.IsCounting = true,
+                    OnPause = () => { RulesetContainer.KeyBindingInputManager.UseParentInput = false; },
+                    OnResume = () => {RulesetContainer.KeyBindingInputManager.UseParentInput = true; },
```
Comment (HoutarouOreki) [inline osu.Game/Screens/Play/Player.cs]: `() => {Rul` typo

Comment (peppy) [reply osu.Game/Screens/Play/Player.cs]: this PR is a wip and this code is not final.

**Review by ekrctb — DISMISSED** [review]
Retry count is not set

**Comment by ekrctb** [comment]
Pressing a key while pausing doesn't increase the key counter after unpausing.

**Comment by peppy** [comment]
Do you mean during the fade period?

**Comment by ekrctb** [comment]
No, key press registers (can click circles) but it doesn't count.
957ed9ca-fd04-4965-9f9a-efae65943ba7

**Comment by peppy** [comment]
Hmm, I can't seem to reproduce this behaviour. Can you see where the logic is failing for this to happen?

**Comment by ekrctb** [comment]
On `IsPaused.ValueChanged`, `KeyBindingInputManager.UseParentInput = !paused;` and it immediately causes input state syncing and fires `KeyDown` -> action press.
But this is before `RulesetContainer.IsPaused.ValueChanged += paused => hudOverlay.KeyCounter.IsCounting = !paused;` this line is executed. So key counter is not count the press.

File: osu.Game/Screens/Play/Player.cs [hunk 2]
```diff
@@ -227,6 +222,11 @@ private void load(AudioManager audio, APIAccess api, OsuConfigManager config)
             hudOverlay.HoldToQuit.Action = Exit;
             hudOverlay.KeyCounter.Visible.BindTo(RulesetContainer.HasReplayLoaded);
 
+            RulesetContainer.IsPaused.BindTo(pauseContainer.IsPaused);
+
+            // schedule to ensure we count any key presses from the current frame (which may affect gameplay).
+            RulesetContainer.IsPaused.ValueChanged += paused => Schedule(() => hudOverlay.KeyCounter.IsCounting = !paused);
```
Comment (ekrctb) [inline osu.Game/Screens/Play/Player.cs]: Hmm, the thing I said is not fixed.
Let me say again. ESC down -> ESC up -> Z down -> ESC down -> you can see key counter is highlighted but the number is not increased.
Using `Schedule` on the another `ValueChanged`, not here, fixes this though I'm not sure if it is the best way.

**Review by ekrctb — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Play/Player.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Screens/Play/Player.cs]: yeah, there has to be a better way to do this. using schedule for inter-class dependencies is a bit ugly. i'll reassess.

Comment (peppy) [reply osu.Game/Screens/Play/Player.cs]: As it turns out, it seems this was not even needed. Could you please re-test? It feels correct at my end.

Comment (ekrctb) [reply osu.Game/Screens/Play/Player.cs]: Seems working fine.

**Review by ekrctb — APPROVED** [review]

**Pull request merged**